File: JXPagerViewExample-OC/JXPagerViewExample-OC/Example/Nest/NestViewController.m
//
//  NestViewController.m
//  JXPagerViewExample-OC
//
//  Created by jiaxin on 2018/10/26.
//  Copyright © 2018 jiaxin. All rights reserved.
//

#import "NestViewController.h"
#import "JXCategoryView.h"
#import "TestNestlistView.h"
#import "JXPagerListRefreshView.h"
#import "PagingViewTableHeaderView.h"
#import "TestNestlistView.h"

static const CGFloat JXTableHeaderViewHeight = 200;
static const CGFloat JXheightForHeaderInSection = 50;

@interface NestViewController () <JXCategoryViewDelegate, JXPagerViewDelegate, JXPagerMainTableViewGestureDelegate>

@property (nonatomic, strong) JXPagerListRefreshView *pagerView;
@property (nonatomic, strong) PagingViewTableHeaderView *userHeaderView;
@property (nonatomic, strong) NSArray <TestNestlistView *> *testListViewArray;
@property (nonatomic, strong) JXCategoryTitleView *categoryView;
@property (nonatomic, strong) NSArray <NSString *> *titles;

@end

@implementation NestViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.title = @"个人中心";
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigationController.navigationBar.translucent = false;
    _titles = @[@"主题一", @"主题二", @"主题三"];

    TestNestlistView *powerListView = [[TestNestlistView alloc] init];
    TestNestlistView *hobbyListView = [[TestNestlistView alloc] init];
    TestNestlistView *partnerListView = [[TestNestlistView alloc] init];

    _testListViewArray = @[powerListView, hobbyListView, partnerListView];

    _userHeaderView = [[PagingViewTableHeaderView alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, JXTableHeaderViewHeight)];

    _categoryView = [[JXCategoryTitleView alloc] initWithFrame:CGRectMake(0, 50, [UIScreen mainScreen].bounds.size.width, 50)];
    self.categoryView.titles = self.titles;
    self.categoryView.backgroundColor = [UIColor greenColor];
    self.categoryView.delegate = self;
    self.categoryView.titleSelectedColor = [UIColor colorWithRed:105/255.0 green:144/255.0 blue:239/255.0 alpha:1];
    self.categoryView.titleColor = [UIColor blackColor];
    self.categoryView.titleColorGradientEnabled = YES;
    self.categoryView.titleLabelZoomEnabled = YES;
    self.categoryView.titleLabelZoomScale = 1.2;

    JXCategoryIndicatorLineView *lineView = [[JXCategoryIndicatorLineView alloc] init];
    lineView.indicatorLineViewColor = [UIColor colorWithRed:105/255.0 green:144/255.0 blue:239/255.0 alpha:1];
    lineView.indicatorLineWidth = 30;
    self.categoryView.indicators = @[lineView];


    //如果不想要下拉刷新的效果，改用JXPagerView类即可
    _pagerView = [[JXPagerListRefreshView alloc] initWithDelegate:self];
    self.pagerView.mainTableView.gestureDelegate = self;
    [self.view addSubview:self.pagerView];

    self.categoryView.contentScrollView = self.pagerView.listContainerView.collectionView;

    self.navigationController.interactivePopGestureRecognizer.enabled = (self.categoryView.selectedIndex == 0);
}


- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];

    self.pagerView.frame = self.view.bounds;
}

#pragma mark - JXPagerViewDelegate

- (UIView *)tableHeaderViewInPagerView:(JXPagerView *)pagerView {
    return self.userHeaderView;
}

- (NSUInteger)tableHeaderViewHeightInPagerView:(JXPagerView *)pagerView {
    return JXTableHeaderViewHeight;
}

- (NSUInteger)heightForPinSectionHeaderInPagerView:(JXPagerView *)pagerView {
    return JXheightForHeaderInSection;
}

- (UIView *)viewForPinSectionHeaderInPagerView:(JXPagerView *)pagerView {
    return self.categoryView;
}

- (NSArray<id<JXPagerViewListViewDelegate>> *)listViewsInPagerView:(JXPagerView *)pagerView {
    return self.testListViewArray;
}

#pragma mark - JXCategoryViewDelegate

- (void)categoryView:(JXCategoryBaseView *)categoryView didSelectedItemAtIndex:(NSInteger)index {
    self.navigationController.interactivePopGestureRecognizer.enabled = (index == 0);
}

-(void)categoryView:(JXCategoryBaseView *)categoryView didClickSelectedItemAtIndex:(NSInteger)index {

}

- (void)categoryView:(JXCategoryBaseView *)categoryView didScrollSelectedItemAtIndex:(NSInteger)index {

}

#pragma mark - JXPagerMainTableViewGestureDelegate

- (BOOL)mainTableViewGestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    if ([self checkIsNestContentScrollView:(UIScrollView *)gestureRecognizer.view] || [self checkIsNestContentScrollView:(UIScrollView *)otherGestureRecognizer.view]) {
        //如果交互的是嵌套的contentScrollView，证明在左右滑动，就不允许同时响应
        return NO;
    }
    return [gestureRecognizer isKindOfClass:[UIPanGestureRecognizer class]] && [otherGestureRecognizer isKindOfClass:[UIPanGestureRecognizer class]];
}

- (BOOL)checkIsNestContentScrollView:(UIScrollView *)scrollView {
    for (TestNestlistView *listView in self.testListViewArray) {
        if (listView.contentScrollView == scrollView) {
            return YES;
        }
    }
    return NO;
}

@end

File: JXPagingView/Example/CategoryNestPaging/NestViewController.swift
//
//  NestViewController.swift
//  JXPagingView
//
//  Created by jiaxin on 2018/11/2.
//  Copyright © 2018 jiaxin. All rights reserved.
//

import UIKit

class NestViewController: UIViewController {
    var naviCategoryView: JXCategoryTitleView!
    var contentScrollView: UIScrollView!
    var pagingVCArray = [BaseViewController]()

    override func viewDidLoad() {
        super.viewDidLoad()

        let titles = ["主题一", "主题二"]
        naviCategoryView = JXCategoryTitleView()
        naviCategoryView.frame = CGRect(x: 0, y: 0, width: 160, height: 30)
        naviCategoryView.titles = titles
        naviCategoryView.cellSpacing = 0
        naviCategoryView.cellWidth = 80
        naviCategoryView.titleColor = .black
        naviCategoryView.titleSelectedColor = .white
        naviCategoryView.isTitleLabelMaskEnabled = true

        let indicatoryView = JXCategoryIndicatorBackgroundView()
        indicatoryView.backgroundViewColor = .red
        indicatoryView.backgroundViewWidth = 80
        indicatoryView.backgroundViewHeight = 30
        indicatoryView.backgroundViewWidthIncrement = 0
        naviCategoryView.indicators = [indicatoryView]

        navigationItem.titleView = naviCategoryView

        contentScrollView = UIScrollView()
        contentScrollView.isPagingEnabled = true
        contentScrollView.bounces = false
        for _ in 0..<titles.count {
            let pagingVC = BaseViewController()
            pagingVC.nestContentScrollView = contentScrollView
            addChild(pagingVC)
            contentScrollView.addSubview(pagingVC.view)
            pagingVCArray.append(pagingVC)
        }
        view.addSubview(contentScrollView)

        naviCategoryView.contentScrollView = contentScrollView
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

        contentScrollView.frame = view.bounds
        contentScrollView.contentSize = CGSize(width: contentScrollView.bounds.size.width*CGFloat(naviCategoryView.titles.count), height: contentScrollView.bounds.size.height)
        contentScrollView.contentOffset.x = CGFloat(naviCategoryView.selectedIndex)*contentScrollView.bounds.size.width
        for (index, vc) in pagingVCArray.enumerated() {
            vc.view.frame = CGRect(x: contentScrollView.bounds.size.width*CGFloat(index), y: 0, width: contentScrollView.bounds.size.width, height: contentScrollView.bounds.size.height)
            vc.pagingView.listContainerView.collectionView.isNestEnabled = true
        }
    }

}
